readthedocs/proxito/middleware.py

```diff
@@ -176,6 +176,66 @@ def add_proxito_headers(self, request, response):
         else:
             response['X-RTD-Version-Method'] = 'path'
 
+    def add_user_headers(self, request, response):
+        """
+        Set specific HTTP headers requested by the user.
+
+        The headers added come from ``projects.models.HTTPHeader`` associated
+        with the ``Domain`` object.
+        """
+        if hasattr(request, 'domain'):
+            response_headers = [header.lower() for header in response]
+            for http_header in request.domain.http_headers.all():
+                if http_header.name.lower() in response_headers:
+                    log.error(
+                        'Overriding an existing response HTTP header. header=%s domain=%s',
+                        http_header.name,
+                        request.domain.domain,
+                    )
+                log.info(
+                    'Adding custtom response HTTP header. header=%s domain=%s',
+                    http_header.name,
+                    request.domain.domain,
+                )
+                response[http_header.name] = http_header.value
+
+    def add_hsts_headers(self, request, response):
+        """
+        Set the Strict-Transport-Security (HSTS) header for docs sites.
+
+        * For the public domain, set the HSTS header if settings.PUBLIC_DOMAIN_USES_HTTPS
+        * For custom domains, check the HSTS values on the Domain object.
+          The domain object should be saved already in request.domain.
+        """
+        if not request.is_secure():
+            # Only set the HSTS header if the request is over HTTPS
+            return response
+
+        host = request.get_host().lower().split(':')[0]
+        public_domain = settings.PUBLIC_DOMAIN.lower().split(':')[0]
+        hsts_header_values = []
+        if settings.PUBLIC_DOMAIN_USES_HTTPS and public_domain in host:
+            hsts_header_values = [
+                'max-age=31536000',
+                'includeSubDomains',
+                'preload',
+            ]
+        elif hasattr(request, 'domain'):
+            domain = request.domain
+            # TODO: migrate Domains with HSTS set using these fields to
+            # ``HTTPHeader`` and remove this chunk of code from here.
```

Comment:
There's only a few of these. A custom Django data migration may be overkill although it would ensure we get it right =).